demos/portal-embed-demo/public/js/portalog/portals-controller.js

```diff
@@ -0,0 +1,239 @@
+            // https://developers.google.com/web/updates/2017/09/autoplay-policy-changes
+            // TODO: autoplay workaround
+        }
+        const messageUI = this._genPortalMessageUI();
+        this.root.appendChild(this.portal);
+        this.root.appendChild(messageUI);
+
+        // Add animation on-click
+        this.portal.addEventListener('click', (evt) => {
+            this.initialY = this.root.getBoundingClientRect().y;
+            this.initialWidth = this.root.getBoundingClientRect().width;
+            this._remove(this.root.querySelector('.message-controller'));
+            this.root.style.transition =
+                `top 0.6s cubic-bezier(.49,.86,.37,1.01),
+                 left 0.3s cubic-bezier(.49,.86,.37,1.01), 
+                width 0.3s cubic-bezier(.49,.86,.37,1.01),
+                padding-top 0.3s cubic-bezier(.49,.86,.37,1.01)`;
+            this.root.style.top = `${this.anmiateY - this.initialY}px`;
+            this.root.style.width = '95%';
+            this.root.style.paddingTop = 'calc(95% * 0.65)';
+            this.root.style.left = 'calc((100% - 95%)/2)';
+            this.portal.postMessage({ control: 'hide' }, this.origin);
+        });
+
+    }
+
+    /**
+     * Adding event listeners
+     */
+    hookEvents() {
+        
+        // Event fires when coming back from TTT Archive
+        window.addEventListener('portalactivate', (evt) => {
+            // reset the overflow style to default
+            document.body.style.overflow = '';
+            // embed predecessor
+            this.embed(evt.adoptPredecessor());
+        });
+
+        // Event fires after the portal on-click animation finishes
+        this.root.addEventListener('transitionend', (evt) => {
+            if (evt.propertyName === 'top') {
+                const isFollowed = document.querySelector('#follow')
+                    .classList.contains('followed');
+
+                // Activate portal with data used in the activated page
+                this.portal.activate({
+                    data: {
+                        followed: isFollowed,
+                        name: 'Yusuke Utsunomiya',
+                        photoSrc: '/img/profile.png',
+                        initialY: this.initialY,
+                        activatedWidth: this.root.getBoundingClientRect().width,
+                        initialWidth: this.initialWidth,
+                    }
+                }).then((_) => {
+                    // Resolves with undefined when adoped as a predecessor
+                    // Check if window.portalHost is present (just in case)
+                    if(!window.portalHost){
+                        return;
+                    }
+
+                    // hide the scroll bar
+                    document.body.style.overflow = 'hidden';
+
+                    // Listen to messages (follow/unfollow)
+                    window.portalHost.addEventListener('message', (evt) => {
```

Comment:
This listener is added every time we swap, so we end up with several copies of this listener. We should either remove it every time we are activated, or track that we have already added it and not do so again.

Reply:
That is a great point!! In my understanding, there is no good way to check if the listener was added to an element (other than adding additional data structure to manage that state), I simply added something like below.

In the constructor...
```javascript
    constructor(embedContainer, src) {
        this.embedContainer = embedContainer;
        this.src = src;
        this.isPortalHostListenerAdded = false;
```
And after activating...
```
// don't add event listeners if it was already added
if(this.isPortalHostListenerAdded) {
    return;
}

// Listen to messages (follow/unfollow)
window.portalHost.addEventListener('message', (evt) => {
    const isFollowed = evt.data.isFollowed;
    this._changeFollowStatus(!isFollowed);
});
this.isPortalHostListenerAdded = true;
```

Do you think this is okay? I kind of feel like there could be more smart way to do this than using dedicated flag but as this is a demo, should be sufficient IMO.

Reply:
sgtm